File: scripts/governance/collect_metrics.py
#!/usr/bin/env python3
"""Governance Tax Tracking — collect metrics for Python + TypeScript codebase"""

import json
import re
from datetime import datetime, timezone
from pathlib import Path


def count_pattern(directory: Path, suffix: str, pattern: str) -> int:
    count = 0
    for f in directory.rglob(f"*{suffix}"):
        if any(p in str(f) for p in ["__pycache__", ".venv", "venv", "node_modules"]):
            continue
        count += len(re.findall(pattern, f.read_text()))
    return count


def count_loc(directory: Path, suffix: str) -> int:
    count = 0
    for f in directory.rglob(f"*{suffix}"):
        if any(p in str(f) for p in ["__pycache__", ".venv", "venv", "node_modules"]):
            continue
        count += len(f.read_text().splitlines())
    return count


metrics = {
    "collected_at": datetime.now(timezone.utc).isoformat(),
    "python_type_holes": count_pattern(Path("app"), ".py", r": Any\b|# type: ignore"),
    "python_noqa": count_pattern(Path("app"), ".py", r"# noqa"),
    "python_silent_catches": 0,  # counted via AST in dedicated script
    "python_clone_count": 0,  # counted via dedicated script
    "ts_any_count": count_pattern(Path("src"), ".ts", r": any\b|as any\b|<any>"),
    "ts_eslint_disable": count_pattern(Path("src"), ".ts", r"eslint-disable"),
    "source_loc_py": count_loc(Path("app"), ".py"),
    "test_loc_py": count_loc(Path("tests"), ".py"),
    "source_loc_ts": count_loc(Path("src"), ".ts"),
    "test_loc_ts": count_loc(Path("tests"), ".ts")
    + count_loc(Path("src/__tests__"), ".ts"),
}

Path(".memory-layer/metrics").mkdir(parents=True, exist_ok=True)
Path(".memory-layer/metrics/latest.json").write_text(json.dumps(metrics, indent=2))

print("Governance metrics collected:")
for k, v in metrics.items():
    if k != "collected_at":
        print(f"  {k}: {v}")

File: scripts/governance/collect-metrics.js
#!/usr/bin/env node
/**
 * Governance Tax Tracking — collect metrics from the current codebase.
 * Writes to .memory-layer/metrics/latest.json
 */

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

function countInFiles(dir, ext, pattern) {
  let count = 0;
  if (!fs.existsSync(dir)) return count;
  function walk(d) {
    for (const f of fs.readdirSync(d)) {
      const full = path.join(d, f);
      const stat = fs.statSync(full);
      if (stat.isDirectory() && !['node_modules', 'dist', '.git', 'coverage', '__pycache__'].includes(f)) {
        walk(full);
      } else if (ext.some(e => full.endsWith(e))) {
        const content = fs.readFileSync(full, 'utf8');
        count += (content.match(pattern) || []).length;
      }
    }
  }
  walk(dir);
  return count;
}

function countLOC(dir, ext) {
  let count = 0;
  if (!fs.existsSync(dir)) return count;
  function walk(d) {
    for (const f of fs.readdirSync(d)) {
      const full = path.join(d, f);
      const stat = fs.statSync(full);
      if (stat.isDirectory() && !['node_modules', 'dist', '.git', 'coverage', '__pycache__'].includes(f)) {
        walk(full);
      } else if (ext.some(e => full.endsWith(e))) {
        count += fs.readFileSync(full, 'utf8').split('\n').length;
      }
    }
  }
  walk(dir);
  return count;
}

const metrics = {
  collected_at: new Date().toISOString(),
  any_count: countInFiles('src', ['.ts', '.tsx'], /: any\b|as any\b|<any>/g),
  eslint_disable: countInFiles('src', ['.ts', '.tsx', '.js'], /eslint-disable/g),
  silent_catches: countInFiles('src', ['.ts', '.tsx', '.js'], /catch\s*(?:\([^)]*\))?\s*\{\s*(?:\/\/[^\n]*)?\s*\}/g),
  source_loc: countLOC('src', ['.ts', '.tsx']),
  test_loc: countLOC('tests', ['.ts', '.tsx', '.py']) + countLOC('src/__tests__', ['.ts', '.tsx']),
  python_type_holes: countInFiles('app', ['.py'], /: Any\b|# type: ignore/g),
  python_noqa: countInFiles('app', ['.py'], /# noqa/g),
};

fs.mkdirSync('.memory-layer/metrics', { recursive: true });
fs.writeFileSync('.memory-layer/metrics/latest.json', JSON.stringify(metrics, null, 2));

console.log('Governance metrics collected:');
Object.entries(metrics).forEach(([k, v]) => {
  if (k !== 'collected_at') console.log(`  ${k}: ${v}`);
});

process.exit(0);
